### PR TITLE
Add check before pthread_join

### PR DESCRIPTION
**Pull request by pvyawaha**
*Description of changes:*
This change adds check so that pthread_join is called only when ota thread is created. 

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

**Comment by aggarw13** [comment]
The set of bug improvements to the OTA demos (across different PRs since the previous release) would be good to add as a single CHANGELOG item point

**Comment by vikivivi** [comment]
* From the manpage of `pthread_join`. See possible errors code upon return.
  
```text
RETURN VALUE
       On success, pthread_join() returns 0; on error, it returns an error number.

ERRORS
       EDEADLK
              A deadlock was detected (e.g., two threads tried to join with each other); or thread specifies the calling thread.

       EINVAL thread is not a joinable thread.

       EINVAL Another thread is already waiting to join with this thread.

       ESRCH  No thread with the ID thread could be found.
```


```C
/**
 * @brief Start OTA demo.
 *
 * @return   EXIT_SUCCESS or EXIT_FAILURE.
 */
static int startOTADemo( void );
```
```C
static int startOTADemo( void )
{
   ...
    if( returnStatus == EXIT_SUCCESS )
    {
        returnStatus = pthread_join( threadHandle, NULL );

        if( returnStatus != EXIT_SUCCESS )
        {
            LogError( ( "Failed to join thread"
                        ",error code = %d",
                        returnStatus ) );
        }
    }

    return returnStatus;
}
```

* Since `startOTADemo()` is documented to return either `EXIT_SUCCESS` or `EXIT_FAILURE`, we should not use error code of `pthread_join()` as the returned error code from `startOTADemo()`. In case someone uses the demo code `startOTADemo()` like this.

```C
if( startOTADemo() != EXIT_FAILURE )   /* Error code of pthread_join will have the same as EXIT_SUCCESS */
{
   ....
}
```

**Comment by pvyawaha** [comment]
>     * From the manpage of `pthread_join`. See possible errors code upon return.
> 
> 
> ```
> RETURN VALUE
>        On success, pthread_join() returns 0; on error, it returns an error number.
> 
> ERRORS
>        EDEADLK
>               A deadlock was detected (e.g., two threads tried to join with each other); or thread specifies the calling thread.
> 
>        EINVAL thread is not a joinable thread.
> 
>        EINVAL Another thread is already waiting to join with this thread.
> 
>        ESRCH  No thread with the ID thread could be found.
> ```
> 
> ```c
> /**
>  * @brief Start OTA demo.
>  *
>  * @return   EXIT_SUCCESS or EXIT_FAILURE.
>  */
> static int startOTADemo( void );
> ```
> 
> ```c
> static int startOTADemo( void )
> {
>    ...
>     if( returnStatus == EXIT_SUCCESS )
>     {
>         returnStatus = pthread_join( threadHandle, NULL );
> 
>         if( returnStatus != EXIT_SUCCESS )
>         {
>             LogError( ( "Failed to join thread"
>                         ",error code = %d",
>                         returnStatus ) );
>         }
>     }
> 
>     return returnStatus;
> }
> ```
> 
>     * Since `startOTADemo()` is documented to return either `EXIT_SUCCESS` or `EXIT_FAILURE`, we should not use error code of `pthread_join()` as the returned error code from `startOTADemo()`. In case someone uses the demo code `startOTADemo()` like this.
> 
> 
> ```c
> if( startOTADemo() != EXIT_FAILURE )   /* Error code of pthread_join will have the same as EXIT_SUCCESS */
> {
>    ....
> }
> ```
> 
>     * Proposed fix at the section of `pthread_join()`.
>       
>       * Declare new variable `returnJoin` to LogError `pthread_join()` error code. Optional if just print failed to join thread.
>       * On success of `pthread_join()` compare return `0`. Just like all other pthread API in OTA demo. (I know EXIT_SUCCESS = 0 as well, but does not belong to the same list of error code)
> 
> 
> ```c
>     if( returnStatus == EXIT_SUCCESS )
>     {
>         if( ( returnJoin = pthread_join( threadHandle, NULL ) ) != 0 )
>         {
>             LogError( ( "Failed to join thread"
>                         ",error code = %d",
>                         returnJoin ) );
>             returnStatus = EXIT_FAILURE;
>         }
>     }
> 
>     return returnStatus;
> ```

Added in commit 
0ad0670

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,7 @@
  - [#1539](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1539) Add retries for MQTT publishing with the backoffAlgorithm library.
  - [#1516](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1516), [#1538](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1538), [#1540](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1540), [#1487](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1487) Minor bugfixes and refactoring of the OTA demos and OTA PAL for POSIX.
  - [#1552](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1552) Update broken links in demo comments and documentation.
+ - [# #1599](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1599) In OTA demo add check before calling pthread_join.
```
Comment (divekarshubham) [inline CHANGELOG.md:23]: There is a new section to the CHANGELOG which documents the changes after the release. https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/main/CHANGELOG.md#main
Could you move this to the above section. Also might change the wording to be consistent with other logs:
"Checking execution status of demo loop before calling pthread_join in OTA demos"

Also might want to include changes of https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1593

Comment (pvyawaha) [reply CHANGELOG.md:31]: Updated in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1599/commits/5fa52c33a0217ce6ec701c5503dc61090f2ccf9c

**Review by divekarshubham — APPROVED** [review]

**Review by n9wxu — APPROVED** [review]

**Comment by vikivivi** [comment]
@pvyawaha Should I raise another issue or you simply make another commit ? You forget to switch LogError to use the new variable `returnJoin`. See my above proposed fix. There is no point to LogError `returnStatus` as it will be `EXIT_SUCCESS`.

```diff
diff --git a/demos/ota/ota_demo_core_http/ota_demo_core_http.c b/demos/ota/ota_demo_core_http/ota_demo_core_http.c
index 5e0a8359..2bbb3cdb 100644
--- a/demos/ota/ota_demo_core_http/ota_demo_core_http.c
+++ b/demos/ota/ota_demo_core_http/ota_demo_core_http.c
@@ -2091,13 +2091,13 @@ static int startOTADemo( void )
         returnJoin = pthread_join( threadHandle, NULL );
 
         if( returnJoin != 0 )
         {
             LogError( ( "Failed to join thread"
                         ",error code = %d",
-                        returnStatus ) );
+                        returnJoin ) );
 
             returnStatus = EXIT_FAILURE;
         }
     }
 
     return returnStatus;
diff --git a/demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c b/demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c
index 4d14ebd5..ed43dca9 100644
--- a/demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c
+++ b/demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c
@@ -1638,13 +1638,13 @@ static int startOTADemo( void )
         returnJoin = pthread_join( threadHandle, NULL );
 
         if( returnJoin != 0 )
         {
             LogError( ( "Failed to join thread"
                         ",error code = %d",
-                        returnStatus ) );
+                        returnJoin ) );
 
             returnStatus = EXIT_FAILURE;
         }
     }
 
     return returnStatus;
```

**Comment by pvyawaha** [comment]
Thank you for pointing it out, I will fix this.

**Pull request merged**